zonemaster/zonemaster-engine
# Ability to run single testcase via test_zone()

## Purpose

Improve the semantics of the test_cases profile property to be more useful and less surprising. This is helpful for
1. people wondering why a test case won't run even though they specified it in their profile, and
2. applications providing flexible ways for users to specify what test cases to run.

## Context

Prerequisite for zonemaster/zonemaster-cli#359.

## Changes

This PR the behavior of Zonemaster::Engine->test_zone() (and its plumbing). It is updated in some important ways:
1. The model for dependencies between test cases is improved. Let's say test case Y has a starting condition that test case X has given a certain signal, but X is disabled in the profile. Before this PR there was no way to run Y without also running X. After this PR disabling X automatically gives the starting signal for Y.
3. Basic01 and Basic02 now respect the test_cases profile property.
4. The Basic test module used to have a special status and was given special treatment in various places. One particular thing the special privilege was used for was to signal that a zone is so broken that no further testing should be done. After this PR there is no discrimination between modules. Any module could in principle signal fatal brokenness — though in practice Basic is the only model that ever does.
5. In some cases when a test case is skipped because its starting condition fails, a special message is emitted. This PR makes it so these messages are only emitted if the skipped test case is enabled in the profile.

This PR also includes a bug fix:
1. A stray TEST_CASE_END message that was emitted from Zonemaster::Engine::Test::DNSSEC->all() is removed.

## How to test this PR

Automated tests:
* A unit test is included ensuring that when you 1) enable any single test case in the profile, and 2) run all enabled test cases, the enabled test case is run exactly once and no other test cases are run.

Manual tests:
* Run `zonemaster-cli --test dnssec` and verify that it doesn't emit stray TEST_CASE_END messages.

Tests that ought to be automated as part of this PR but that I'm lacking test data for:
* Verify that enabled test cases can disable dependent test cases. Here I've compiled a list of all signals involved in dependencies between test cases. Each entry consists of three sets. The first one is a set of signaling test cases, the second is a set of test cases with a starting condition, the third is a set of messages involved in the signal. I don't have test zones for any of these, but if you have some I'll add unit tests for them.
  1. ({basic01}, {basic02, basic03}, {B01_CHILD_FOUND})
  2. ({basic02}, {basic03}, {B02_AUTH_RESPONSE_SOA})
  3. ({basic02}, *all test cases in all test modules other than Basic*, {B02_NO_DELEGATION, B02_AUTH_RESPONSE_SOA, *a somewhat external "is-undelegated" property*})
  4. ({address02}, {address03}, {NAMESERVERS_IP_WITH_REVERSE})
  6. ({dnssec07}, *all other test cases in the DNSSEC test module*, {NEITHER_DNSKEY_NOR_DS})
  7. ({dnssec07}, {dnssec06}, {DNSKEY_BUT_NOT_DS, DNSKEY_AND_DS})
  8. ({syntax01}, {syntax04, syntax05, syntax06, syntax07, syntax08}, {ONLY_ALLOWED_CHARS})
  9. ({syntax05}, {syntax06, syntax07}, {NO_RESPONSE_SOA_QUERY})
  10. ({zone08}, {zone09}, {NO_RESPONSE_MX_QUERY})
  11. ({zone02, zone03, zone04, zone05, zone06, zone07}, {zone10, zone11}, {NO_RESPONSE_SOA_QUERY})
* Verify that the special skipping messages are emitted if and only if their associated test case is both skipped and enabled in the profile. Here I've compiled a list of these special messages. Each entry consists of a pair of a test case and a special message. I don't have test zones for any of these, but if you have some I'll add unit tests for them.
  1. (basic03, HAS_NAMESERVER_NO_WWW_A_TEST)
  2. (dnssec06, ADDITIONAL_DNSKEY_SKIPPED)
